### PR TITLE
Migrate existing adapter projects

### PR DESCRIPTION
**Pull request by UncleSamSwiss**
**PR Checklist:**

-   [x] Provide a meaningful description to this PR or mention which issues this fixes.
-   [x] Add tests for your change. This includes negative tests (i.e. inputs that need to fail) as well as baseline tests (i.e. how should the directory structure look like?).
-   [x] Run the test suite with `npm test`
-   [ ] ~~If there are baseline changes, review them and make a separate commit for them with the comment "accept baselines" if they are desired changes~~
-   [x] Ensure the project builds with `npm run build`
-   [ ] ~~If you added a required option, also add it to the template creation (`.github/create_templates.ts`)~~
-   [ ] ~~Add a detailed migration description to `docs/updates` explaining what the user needs to do when manually updating an existing project~~
-   [x] Add your changes to `CHANGELOG.md` (referencing the migration description and this PR or the issue you fixed)

**Description:**  
Allow to run the adapter creator with the answers pre-filled from an existing adapter directory (the given path needs to point to the adapter base directory where `io-package.json` is found). It will try to pre-fill all answers, but the user still has to acknowledge all answers.

### CODE REVIEW

File: src/lib/questions.ts [hunk 1]
```diff
@@ -416,6 +469,12 @@ export const questionsAndText: (
 				hint: "(To some device or some service)",
 				initial: "no",
 				choices: ["yes", "no"],
+				import: (ctx) =>
+					ctx.ioPackageJson.instanceObjects?.find(
```
Comment (AlCalzone) [inline src/lib/questions.ts]:
```suggestion
					ctx.ioPackageJson.instanceObjects?.some(
```

File: src/lib/questions.ts [hunk 2]
```diff
@@ -435,6 +494,10 @@ export const questionsAndText: (
 				message:
 					"Which language do you want to use to code the adapter?",
 				choices: ["JavaScript", "TypeScript"],
+				import: async (ctx) =>
+					(await ctx.hasFilesWithExtension("src", ".ts"))
```
Comment (AlCalzone) [inline src/lib/questions.ts]: This also finds `.d.ts` files, which might exist in JS projects

File: src/lib/questions.ts [hunk 3]
```diff
@@ -443,6 +506,8 @@ export const questionsAndText: (
 				message: "Use React for the Admin UI?",
 				initial: "no",
 				choices: ["yes", "no"],
+				import: async (ctx) =>
+					(await ctx.directoryExists("admin/src")) ? "yes" : "no",
```
Comment (AlCalzone) [inline src/lib/questions.ts]: I'd check for existence of JSX or TSX files instead (not called tab.jsx/tsx)

File: src/lib/questions.ts [hunk 4]
```diff
@@ -496,6 +585,8 @@ export const questionsAndText: (
 				message: "Do you prefer tab or space indentation?",
 				initial: "Tab",
 				choices: ["Tab", "Space (4)"],
+				import: async (ctx) =>
+					(await ctx.analyzeCode("\t", "  ")) ? "Tab" : "Space (4)",
```
Comment (AlCalzone) [inline src/lib/questions.ts:604]: If ESLint is used, we could look at the indent or @typescript-eslint/indent rules instead.

File: src/lib/questions.ts [hunk 5]
```diff
@@ -504,6 +595,8 @@ export const questionsAndText: (
 				message: "Do you prefer double or single quotes?",
 				initial: "double",
 				choices: ["double", "single"],
+				import: async (ctx) =>
+					(await ctx.analyzeCode('"', "'")) ? "double" : "single",
```
Comment (AlCalzone) [inline src/lib/questions.ts:614]: If ESLint is used, we could look at the quotes rule instead.

File: src/lib/questions.ts [hunk 6]
```diff
@@ -535,19 +632,26 @@ export const questionsAndText: (
...
 				type: "input",
 				name: "authorGithub",
 				message: "What's your name/org on GitHub?",
 				initial: ((answers: Answers) => answers.authorName) as any,
 				action: checkAuthorName,
+				import: (ctx) =>
+					ctx.ioPackageJson.common?.extIcon?.replace(
+						/^.+?\.com\/([^\/]+)\/.+$/,
+						"$1",
+					),
```
Comment (AlCalzone) [inline src/lib/questions.ts]: I'd look at package.json -> author.name instead

File: src/cli.ts [hunk 7]
```diff
@@ -76,12 +85,38 @@ async function ask(): Promise<Answers> {
...
+
 	async function askQuestion(q: Question): Promise<void> {
 		if (testCondition(q.condition, answers)) {
 			// Make properties dependent on previous answers
 			if (typeof q.initial === "function") {
 				q.initial = q.initial(answers);
 			}
+			if (importContext && q.import) {
+				let imported = q.import(importContext, answers, q);
+				if (imported && typeof (imported as any).then === "function") {
```
Comment (AlCalzone) [inline src/cli.ts]:
```suggestion
				if (imported instanceof Promise) {
```
maybe?

File: src/lib/importContext.ts [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+			!!(await readdir(path.join(this.baseDir, dirName))).find((f) =>
+				f.toLowerCase().endsWith(extension.toLowerCase()),
+			)
+		);
+	}
+
+	public hasDevDependency(packageName: string): boolean {
+		return (
+			this.packageJson.devDependencies &&
+			this.packageJson.devDependencies.hasOwnProperty(packageName)
```
Comment (AlCalzone) [inline src/lib/importContext.ts]:
```suggestion
			this.packageJson.devDependencies?.hasOwnProperty(packageName)
```

File: src/lib/importContext.ts [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+		try {
+			const tsMain = path.join(
+				this.baseDir,
+				"src",
+				this.packageJson.main.replace(/\.js$/, ".ts"),
+			);
+			if (existsSync(tsMain)) {
+				// most probably TypeScript
+				return await readFile(tsMain, { encoding: "utf8" });
```
Comment (AlCalzone) [inline src/lib/importContext.ts]:
```suggestion
				return readFile(tsMain, { encoding: "utf8" });
```

File: src/lib/importContext.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+			} else {
+				return await readFile(
```
Comment (AlCalzone) [inline src/lib/importContext.ts]:
```suggestion
				return readFile(
```

File: src/lib/questions.ts (hunk 4 above)
Comment (AlCalzone) [reply src/lib/questions.ts:604]: Also, shouldn't that be 4 spaces instead?

**Review by AlCalzone — CHANGES_REQUESTED** [review]

File: src/lib/questions.ts (hunk 4 above)
Comment (UncleSamSwiss) [reply src/lib/questions.ts:604]: First comment: yes, but in any case this comparison will work - parsing rules is a bit more tricky and might change over time.
Second comment: true, but I prefer 2 spaces, so we kind of "upgrade" from 2 spaces to 4 but keep spaces. Two spaces will also match four spaces ;-).

File: src/lib/questions.ts (hunk 6 above)
Comment (UncleSamSwiss) [reply src/lib/questions.ts]: I thought so too, but some old projects might not have this in package.json. In addition, for adapters inside an org, that will not work. Example: https://github.com/iobroker-community-adapters/ioBroker.squeezebox/blob/5c7b6aa9181393d57cf13fa1d4d39c1f17904b7b/package.json#L5-L7
According to the adapter tester, `common.extIcon` must match the right repo.

File: src/lib/questions.ts (hunk 5 above)
Comment (UncleSamSwiss) [reply src/lib/questions.ts:614]: See above.

File: src/cli.ts (hunk 7 above)
Comment (UncleSamSwiss) [reply src/cli.ts]: Thanks, was looking for that but apparently used the wrong Google search terms ;-)

File: src/lib/importContext.ts (hunk 10 above)
Comment (UncleSamSwiss) [reply src/lib/importContext.ts]: readFile is async, or why would you remove the await?

File: src/lib/importContext.ts (hunk 9 above)
Comment (UncleSamSwiss) [reply src/lib/importContext.ts]: readFile is async, or why would you remove the await?

Comment (AlCalzone) [reply src/lib/importContext.ts]: `return await` is an anti-pattern in many cases - it creates an additional Promise that just returns the result of the `await`ed function.
In an async function, either `await` something and discard the result or `return` the Promise directly.

https://eslint.org/docs/rules/no-return-await

File: src/lib/importContext.ts (hunk 10 above)
Comment (AlCalzone) [reply src/lib/importContext.ts]: see other comment

**Comment by AlCalzone** [comment]
Your tests are failing.

**Review by AlCalzone — APPROVED** [review]

**Pull request merged**